tools/rpkg/R/relational.R

```diff
@@ -151,6 +153,30 @@ rel_aggregate <- rapi_rel_aggregate
 #' rel2 <- rel_order(rel, list(expr_reference("hp")))
 rel_order <- rapi_rel_order
 
+
+
+expr_window <- function(window_function, partitions=list(),
+				window_boundary_start="unbounded_preceding",
+				window_boundary_end="current_row_range",
+				start_expr = list(), end_expr=list(), offset_expr=list(), default_expr=list()) {
+  expr_window_(window_function, partitions, tolower(window_boundary_start), tolower(window_boundary_end), start_expr, end_expr, offset_expr, default_expr)
+}
+
+window_boundaries <- c("unbounded_preceding",
+                       "unbounded_following",
+                       "current_row_range",
+                       "current_row_rows",
+                       "expr_preceding_rows",
+                       "expr_following_rows",
+                       "expr_preceding_range")
+
+expr_window_ <- function (window_function, partitions=list(), window_boundary_start=window_boundaries,
+          window_boundary_end=window_boundaries, start_expr = list(), end_expr=list(), offset_expr=list(), default_expr=list()) {
```

Comment:
Dumb question... are expressions lists? I ask because the start/end/offset/default arguments are optional functions.

Reply:
so initially I passed these expressions as lists because they were optional, and you can't pass an external null pointer to the relational api directly from R. The next easiest thing was just to pass a list. 

I've since modified it so you can call into the relational api to get an external_pointer that points to null_ptr, and optional arguments can just pass the output of calling `sexp_null_ptr()`.

Reply:
We might want to patch cpp11 (again). We're patching it anyway.

The `external_pointer<>` template can be constructed to contain a `NULL` pointer, but only with the default constructor. In the below example, the commented C++ code fails, but should work IMO. This would simplify the wrapping code a lot.

``` r
library(cpp11)

cpp_source(
  code = '
  #include "cpp11/external_pointer.hpp"
  #include "cpp11/strings.hpp"

  [[cpp11::register]]
  cpp11::external_pointer<int> test() {
    // return cpp11::external_pointer<int>(R_NilValue);
    return cpp11::external_pointer<int>(nullptr);
  }
  '
)

test()
#> NULL
```

<sup>Created on 2023-03-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
This patching can be a separate PR as well right? Would like to avoid scope creep on this PR